### ISSUE TITLE
heap-use-after-free in set_bufref

### ISSUE THREAD

**Issue by gy741** [vim/vim#2448]
Hello.

I found a heap-use-after-free bug in vim.

Please confirm.

Thanks.

Summary: heap-use-after-free
OS: CentOS 7 64bit
Version: commit 4697399e8c805325009351a2488e3da530d0af38
PoC Download:  [set_bufref.zip](https://github.com/vim/vim/files/1561152/set_bufref.zip)

 Steps to reproduce:
 1.Download the .POC files.
 2.Compile the source code with ASan.
 3.Execute the following command
   : ./vim -u NONE -Z -X -e -s -S $FILE -c :qa!

```
=================================================================
==16991==ERROR: AddressSanitizer: heap-use-after-free on address 0x6240000261b8 at pc 0x0000004fc86d bp 0x7fffa0f581e0 sp 0x7fffa0f581d8
READ of size 4 at 0x6240000261b8 thread T0
    #0 0x4fc86c in set_bufref /home/karas/vim/src/buffer.c:375:46
    #1 0xecf0d5 in load_dummy_buffer /home/karas/vim/src/quickfix.c:4615:3
    #2 0xebf861 in ex_vimgrep /home/karas/vim/src/quickfix.c:4312:12
    #3 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #4 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #5 0x813c17 in do_source /home/karas/vim/src/ex_cmds2.c:4398:5
    #6 0x8100d7 in cmd_source /home/karas/vim/src/ex_cmds2.c:4011:14
    #7 0x8101f6 in ex_source /home/karas/vim/src/ex_cmds2.c:3986:2
    #8 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #9 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #10 0x82c495 in do_cmdline_cmd /home/karas/vim/src/ex_docmd.c:671:12
    #11 0x1654d44 in exe_commands /home/karas/vim/src/main.c:2948:2
    #12 0x164e8e6 in vim_main2 /home/karas/vim/src/main.c:800:2
    #13 0x163fa4d in main /home/karas/vim/src/main.c:429:12
    #14 0x7f5d0824182f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
    #15 0x41aaa8 in _start (/home/karas/vim/src/vim+0x41aaa8)

0x6240000261b8 is located 184 bytes inside of 7216-byte region [0x624000026100,0x624000027d30)
freed by thread T0 here:
    #0 0x4baa50 in __interceptor_cfree.localalias.0 (/home/karas/vim/src/vim+0x4baa50)
    #1 0xbf013b in vim_free /home/karas/vim/src/misc2.c:1801:2
    #2 0x9bc106 in apply_autocmds_group /home/karas/vim/src/fileio.c:9761:6
    #3 0x97dbbf in apply_autocmds_exarg /home/karas/vim/src/fileio.c:9268:12
    #4 0x968114 in readfile /home/karas/vim/src/fileio.c:389:10
    #5 0xeceee8 in load_dummy_buffer /home/karas/vim/src/quickfix.c:4602:6
    #6 0xebf861 in ex_vimgrep /home/karas/vim/src/quickfix.c:4312:12
    #7 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #8 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #9 0x813c17 in do_source /home/karas/vim/src/ex_cmds2.c:4398:5
    #10 0x8100d7 in cmd_source /home/karas/vim/src/ex_cmds2.c:4011:14
    #11 0x8101f6 in ex_source /home/karas/vim/src/ex_cmds2.c:3986:2
    #12 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #13 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #14 0x82c495 in do_cmdline_cmd /home/karas/vim/src/ex_docmd.c:671:12
    #15 0x1654d44 in exe_commands /home/karas/vim/src/main.c:2948:2
    #16 0x164e8e6 in vim_main2 /home/karas/vim/src/main.c:800:2
    #17 0x163fa4d in main /home/karas/vim/src/main.c:429:12
    #18 0x7f5d0824182f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291

previously allocated by thread T0 here:
    #0 0x4babd8 in malloc (/home/karas/vim/src/vim+0x4babd8)
    #1 0xbec3b8 in lalloc /home/karas/vim/src/misc2.c:954:21
    #2 0xbec7d7 in alloc_clear /home/karas/vim/src/misc2.c:876:9
    #3 0x50a92b in buflist_new /home/karas/vim/src/buffer.c:2006:17
    #4 0xecec46 in load_dummy_buffer /home/karas/vim/src/quickfix.c:4577:14
    #5 0xebf861 in ex_vimgrep /home/karas/vim/src/quickfix.c:4312:12
    #6 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #7 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #8 0x813c17 in do_source /home/karas/vim/src/ex_cmds2.c:4398:5
    #9 0x8100d7 in cmd_source /home/karas/vim/src/ex_cmds2.c:4011:14
    #10 0x8101f6 in ex_source /home/karas/vim/src/ex_cmds2.c:3986:2
    #11 0x847a01 in do_one_cmd /home/karas/vim/src/ex_docmd.c:2908:2
    #12 0x826edd in do_cmdline /home/karas/vim/src/ex_docmd.c:1071:17
    #13 0x82c495 in do_cmdline_cmd /home/karas/vim/src/ex_docmd.c:671:12
    #14 0x1654d44 in exe_commands /home/karas/vim/src/main.c:2948:2
    #15 0x164e8e6 in vim_main2 /home/karas/vim/src/main.c:800:2
    #16 0x163fa4d in main /home/karas/vim/src/main.c:429:12
    #17 0x7f5d0824182f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291

SUMMARY: AddressSanitizer: heap-use-after-free /home/karas/vim/src/buffer.c:375:46 in set_bufref
Shadow bytes around the buggy address:
  0x0c487fffcbe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c487fffcbf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c487fffcc00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c487fffcc10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c487fffcc20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c487fffcc30: fd fd fd fd fd fd fd[fd]fd fd fd fd fd fd fd fd
  0x0c487fffcc40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c487fffcc50: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c487fffcc60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c487fffcc70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c487fffcc80: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==16991==ABORTING
```
   
=================================================================
[Acknowledgement]
This work was supported by ICT R&D program of MSIP/IITP. [R7518-16-1001, Innovation hub for high Performance Computing]

**Comment by dpelle** [comment]
I can reproduce it with this simplified case:

```
$ vim -u NONE -c'au * x bw' -c 'lv½ /x'
Vim: Caught deadly signal SEGV
```

```
$ valgrind ./vim -u NONE -e -s -c'au * x bw' -c 'lv½ /x'
==30479== Memcheck, a memory error detector
==30479== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==30479== Using Valgrind-3.14.0.GIT and LibVEX; rerun with -h for copyright info
==30479== Command: ./vim -u NONE -e -s -cau\ *\ x\ bw -c lv½\ /x
==30479== 
==30479== Invalid read of size 4
==30479==    at 0x40FE78: set_bufref (buffer.c:375)
==30479==    by 0x51F22C: load_dummy_buffer (quickfix.c:4615)
==30479==    by 0x51F22C: ex_vimgrep (quickfix.c:4312)
==30479==    by 0x4689BC: do_one_cmd (ex_docmd.c:2908)
==30479==    by 0x464C3D: do_cmdline (ex_docmd.c:1071)
==30479==    by 0x62586C: exe_commands (main.c:2948)
==30479==    by 0x62586C: vim_main2 (main.c:800)
==30479==    by 0x624414: main (main.c:429)
==30479==  Address 0xce1da38 is 184 bytes inside a block of size 7,232 free'd
==30479==    at 0x4C2ECF0: free (vg_replace_malloc.c:530)
==30479==    by 0x492EF7: apply_autocmds_group (fileio.c:9772)
==30479==    by 0x4868F0: readfile (fileio.c:0)
==30479==    by 0x51F1F0: load_dummy_buffer (quickfix.c:4602)
==30479==    by 0x51F1F0: ex_vimgrep (quickfix.c:4312)
==30479==    by 0x4689BC: do_one_cmd (ex_docmd.c:2908)
==30479==    by 0x464C3D: do_cmdline (ex_docmd.c:1071)
==30479==    by 0x62586C: exe_commands (main.c:2948)
==30479==    by 0x62586C: vim_main2 (main.c:800)
==30479==    by 0x624414: main (main.c:429)
==30479==  Block was alloc'd at
==30479==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)
==30479==    by 0x4D4D87: lalloc (misc2.c:954)
==30479==    by 0x4D4EE3: alloc_clear (misc2.c:876)
==30479==    by 0x4115FA: buflist_new (buffer.c:2006)
==30479==    by 0x51F129: load_dummy_buffer (quickfix.c:4577)
==30479==    by 0x51F129: ex_vimgrep (quickfix.c:4312)
==30479==    by 0x4689BC: do_one_cmd (ex_docmd.c:2908)
==30479==    by 0x464C3D: do_cmdline (ex_docmd.c:1071)
==30479==    by 0x62586C: exe_commands (main.c:2948)
==30479==    by 0x62586C: vim_main2 (main.c:800)
==30479==    by 0x624414: main (main.c:429)
...snip many other errors....
```
Doing :bwipeout inside an autocommand is asking for trouble of
course but Vim should ideally not crash.

**Comment by brammool** [comment]
Fixed by  8.0.1403
